prov/efa/src/efa_hmem.c

```diff
@@ -298,6 +298,49 @@ static int efa_hmem_info_init_synapseai(struct efa_hmem_info *synapseai_info, st
 	return 0;
 }
 
+/**
+ * @brief  Validate an FI_OPT_FI_HMEM_P2P (FI_OPT_ENDPOINT) option for a
+ *         specified HMEM interface
+ *
+ * @param  domain  The efa_domain struct which contains an efa_hmem_info array
+ * @param  iface   The fi_hmem_iface enum of the HMEM interface to validate
+ * @param  p2p_opt The P2P option to validate
+ *
+ * @return FI_SUCCESS if the P2P option is valid for the given interface
+ *         FI_OPNOTSUPP if the P2P option is invalid
+ *         FI_ENODATA if the given HMEM interface was not initialized
+ *         FI_EINVAL if p2p_opt is not a valid FI_OPT_FI_HMEM_P2P option
+ */
+int efa_hmem_validate_p2p_opt(struct efa_domain *efa_domain, enum fi_hmem_iface iface, int p2p_opt)
+{
+	struct efa_hmem_info info = efa_domain->hmem_info[iface];
+
+	if (OFI_UNLIKELY(!info.initialized))
+		return FI_ENODATA;
+	
+	switch (p2p_opt) {
+
+	/* REQUIRED is valid even if an interface doesn't require P2P */
+	case FI_HMEM_P2P_REQUIRED:
+		return info.p2p_supported ? FI_SUCCESS : FI_EOPNOTSUPP;
+
+	/*
+	 * Invalidate ENABLED or PREFERRED if an iface requires P2P, or if an iface
+	 * doesn't support P2P at all
+	 */
+	case FI_HMEM_P2P_PREFERRED:
+	case FI_HMEM_P2P_ENABLED:
+		return info.p2p_supported && !info.p2p_required
+			? FI_SUCCESS
+			: FI_EOPNOTSUPP;
+
+	/* DISABLED is valid unless an interface requires P2P */
+	case FI_HMEM_P2P_DISABLED:
+		return !info.p2p_required ? FI_SUCCESS : FI_EOPNOTSUPP;
+	}
```

Comment:
I think the flag name `p2p_required` is a bit confusing.

The flag introduced here meant to say whether p2p is required by the implementation (which user cannot change).

The confusing part is that there is setopt flag p2p_required, which means that even if provider can use non-p2p to support the memory type, the user does not want provider to do that.

Reply:
@lrbison `FI_HMEM_P2P_DISABLED` should be supported when non-p2p `FI_HMEM_CUDA` modes are supported; the switch case here will be necessary then.

@wzamazon I can see how the nomenclature of `p2p_required` might be confusing since it might indicate a user-defined restriction for libfabric. Perhaps something like `p2p_only`?

Reply:
I think a different name is needed. But neither `p2p_only` or `p2p_mandatory` sounds right. should be something more explicit like `p2p_required_by_implementation`.